### PR TITLE
ipc: icmsg: Dedicated workqueue for ICMSG backend

### PR DESCRIPTION
**Pull request by e-rk**
The ICMSG backend now has a dedicated workqueue to process incoming IPC messages. The system workqueue is no longer utilized for that purpose.

Testing shows that in certain scenarios substituting a RPMsg backend with ICMsg results in deadlocks.
The deadlocks were a symptom of running a synchronous RPC protocol from the context of the system workqueue and transpired as follows:

1. The RPC protocol sends a request over the ICMsg backend on the system workqueue thread.
2. The RPC protocol puts the thread to sleep until response is received. This puts the system workqueue thread to sleep.
3. The response to the request arrives over ICMsg backend.
4. The backend signals a work item to the system workqueue.
5. The system workqueue is unable to process the response due to being previously pended on the RPC request.

The deadlock was initially observed with the nrf-802154 driver in conjuntion with the IPv6 stack.

To prevent this condition from occurring, the approach was selected to give ICMsg a dedicated workqueue thread.

### CODE REVIEW

**Comment by hubertmis** [comment]
I'm not convinced. I think the problem is:

> 2. The RPC protocol puts the thread to sleep until response is received. This **puts the system workqueue thread to sleep**.

The icmsg library is intended to be lightweight, not requiring dependencies like additional workqueues.

**Comment by emob-nordic** [comment]
> I'm not convinced. I think the problem is:
> 
> > 2. The RPC protocol puts the thread to sleep until response is received. This **puts the system workqueue thread to sleep**.
> 
> The icmsg library is intended to be lightweight, not requiring dependencies like additional workqueues.
> 
> Maybe if the RPC protocol must put its thread to sleep it should use a dedicated workqueue instead of the system one?

I agree with @hubertmis . The issue is related to they way you use icmsg.

**Comment by e-rk** [comment]
@hubertmis @emob-nordic I think the problem is more fundamental and can't be solved by simply giving the RPC protocol its own thread.
Taking nrf-802154 as an example, its RPC protocol is by design fully synchronous. It emulates calling a remote function as if it were a local one. That means:
1. Sending a request to send the function arguments
2. Waiting for response to give back the return value

Let's say the system workqueue thread wants to send a frame. When the request is sent, the thread must not continue to run because doing so would require giving the caller some garbage return value. Delegating RPC processing to another thread can't help here, because we still won't be able to provide the return value.

If having a dedicated workqueue for ICMsg is not acceptable, then this means that synchronous RPC protocols must never be used from the system workqueue context. That will have major ramifications not only for nrf-802154, but for other protocols that utilize synchronous request (for example `hci.h:bt_hci_cmd_send_sync` or `nrf_rpc.h:nrf_rpc_cbor_cmd_rsp`.
The users of those functions (BLE stack, network stack, openthread and everything above those modules) would have to be extra careful to never call those functions from the system workqueue.

Perhaps a more acceptable approach would be having a dedicated workqueue as a Kconfig option? The Kconfig would be `y` by default (dedicated workqueue enabled) and users that are able to meet the requirement of never doing synchronous calls from system workqueue context would be able to disable the dedicated thread.

**Review by hubertmis — CHANGES_REQUESTED** [review]
> then this means that synchronous RPC protocols must never be used from the system workqueue context

The reason that synchronous RPC protocols should not use the system workqueue is not the icmsg design, but the Zephyr system workqueue usage guidelines:

https://github.com/zephyrproject-rtos/zephyr/blob/ad7ae7cc1248781f354b1a86afb036c581e2dbb9/doc/kernel/services/threads/workqueue.rst?plain=1#L215

> A new workqueue can be justified if it is not possible for its work items to co-exist with existing system workqueue work items without an unacceptable impact; for example, **if the new work items perform blocking operations** that would delay other system workqueue processing to an unacceptable degree.

**Comment by emob-nordic** [comment]
> Taking nrf-802154 as an example, its RPC protocol is by design fully synchronous. It emulates calling a remote function as if it were a local one. That means:
> 
> 1. Sending a request to send the function arguments
> 2. Waiting for response to give back the return value

I don't think that rpc being full synchronous give authority to block the whole system by putting to sleep system work queue. From rpc side any function call is treated as local one and that's fair enought, but in fact this function is not local and rpc is aware of that. Blocking system in such case should not happen (imho).

**Review by hubertmis — COMMENTED** [review]
After more discussion, I recently realized that icmsg calls user-defined callbacks from the workqueue context. The icmsg has no control if this callback is blocking or not. Because of this, by default icmsg should use a dedicated workqueue like proposed in this PR.

However, I think we should also implement this "more acceptable" approach suggested by @e-rk: if icmsg user can guarantee that none of the registered callbacks are blocking, the user can select in Kconfig that icmsg uses the system workqueue.

Not approving yet. Let's discuss first if the "more acceptable" option should be included in this PR, or in the future.

FYI @rakons

File: subsys/ipc/ipc_service/lib/Kconfig.icmsg
Comment (hubertmis) [inline subsys/ipc/ipc_service/lib/Kconfig.icmsg:46]: Please add a note that user-defined callbacks are called from the selected workqueue, so the user is responsible to obey system workqueue rules when this option is disabled.

**Review by hubertmis — DISMISSED** [review]
Looks good. I think it would be good to clarify the implications of disabling the dedicated workqueue. But it's not a blocker for me.

I think @emob-nordic or @rakons should approve this before merging

File: subsys/ipc/ipc_service/lib/Kconfig.icmsg
Comment (e-rk) [reply subsys/ipc/ipc_service/lib/Kconfig.icmsg:46]: Added the note.

File: subsys/ipc/ipc_service/lib/Kconfig.icmsg
Comment (rakons) [inline subsys/ipc/ipc_service/lib/Kconfig.icmsg:40]: Can we make it disabled by default? The issue is we are not sure if everything in the system is compatible with this configuration. I am not saying it is not - but we did not check that yet.

File: subsys/ipc/ipc_service/lib/Kconfig.icmsg
Comment (rakons) [inline subsys/ipc/ipc_service/lib/Kconfig.icmsg:59]: Leaving dedicated WQ enabled by default is risky - from the top of my mind some of the applications are increasing system workque stack to fit. Now such configurations may fail.

File: subsys/ipc/ipc_service/lib/icmsg.c
Comment (rakons) [inline subsys/ipc/ipc_service/lib/icmsg.c:35]: static struct k_work_q *const workq = ...

**Review by rakons — CHANGES_REQUESTED** [review]

File: subsys/ipc/ipc_service/lib/Kconfig.icmsg
Comment (e-rk) [reply subsys/ipc/ipc_service/lib/Kconfig.icmsg:40]: Enabling this by default makes the backend compatible in usage scenarios where RPMsg has been currently used. From that point of view, this approach follows the principle of least surprise.

File: subsys/ipc/ipc_service/lib/Kconfig.icmsg
Comment (e-rk) [reply subsys/ipc/ipc_service/lib/Kconfig.icmsg:59]: I don't know if I understand the problem. The defconfig for a particular platform/CPU should set the stack size to be large enough to strike the balance between the most common use-cases and memory utilization.

File: subsys/ipc/ipc_service/lib/icmsg.c
Comment (e-rk) [reply subsys/ipc/ipc_service/lib/icmsg.c]: Fixed.

File: subsys/ipc/ipc_service/lib/Kconfig.icmsg
Comment (emob-nordic) [inline subsys/ipc/ipc_service/lib/Kconfig.icmsg:38]:
``` suggestion
config IPC_SERVICE_BACKEND_ICMSG_WQ_ENABLE
```

**Review by eobalski — APPROVED** [review]
> After more discussion, I recently realized that icmsg calls user-defined callbacks from the workqueue context. The icmsg has no control if this callback is blocking or not. Because of this, by default icmsg should use a dedicated workqueue like proposed in this PR.

I can accept this PR as a workaround but not a proper fix. This solution can not be levered, because we missed one sentence in the ipc_service documentation. Which is:
`User registered IPC service callbacks shall not be blocking. The received data should be processed asap or in another thread`. That's the only thing that is missing in the documentation which stops me from blocking this PR in official way. I believe you agree with me.

I would not persist in blocking this as it has clear justification now, but please note this is thin ice and sooner or later we would have to fix the way system workq is overused in Zephyr.

Please correct this small typo.

**Comment by emob-nordic** [comment]
I recently was considering if data received could be processed directly from mbox interrupt, but did not come to any conclusions yet (little time).

File: subsys/ipc/ipc_service/lib/Kconfig.icmsg
Comment (e-rk) [reply subsys/ipc/ipc_service/lib/Kconfig.icmsg:38]: Fixed

**Review by rakons — APPROVED** [review]

**Comment by emob-nordic** [comment]
@carlocaione  could you have a look?

**Review by hubertmis — APPROVED** [review]

File: subsys/ipc/ipc_service/lib/icmsg.c
Comment (carlocaione) [inline subsys/ipc/ipc_service/lib/icmsg.c:534]: just a curiosity: why this is enabled by `SYS_INIT` and not as part of the init backend?

Comment (e-rk) [reply subsys/ipc/ipc_service/lib/icmsg.c:534]: There is a single workqueue shared among all backend instances. If you mean why not in `icmsg_open`, that would result in multiple initializations without some static flag. If you think a static flag is preferable, then I can change it to that.

Comment (carlocaione) [reply subsys/ipc/ipc_service/lib/icmsg.c:534]: I see, not a problem but `SYS_INIT` is being currently reworked by @gmarull and I'm not sure what's in the future for that. @gmarull any opinion on this?

Comment (gmarull) [reply subsys/ipc/ipc_service/lib/icmsg.c:534]: SYS_INIT won't disappear anytime soon, so feel free to use it. In any case, if SYS_INIT is ever deprecated a replacement will exist.

**Review by carlescufi — APPROVED** [review]

**Pull request merged**
